### PR TITLE
Phragmen solution should submit for current era and be checked against current era

### PR DESCRIPTION
**Pull request by thiolliere**
phragmen solution should be submitted for the current era. the pallet-session can decide to delay by one or multiple era and dynamically. The active era should only be used for rewards stuff.

Phragmen solution submittion submit solution for the next current era so they should be submitted in the current_era, regardless of the active_era.

### CODE REVIEW

File: frame/staking/src/tests.rs [hunk 1]
```diff
@@ -3006,7 +3060,7 @@ mod offchain_phragmen {
 					winners,
 					compact,
 					score,
-					active_era(),
+					current_era(),
```
Comment (thiolliere) [inline frame/staking/src/tests.rs:3063]: note that for all those test current_era == active_era anyway.

File: frame/staking/src/tests.rs [hunk 2]
```diff
@@ -2990,6 +2990,60 @@ mod offchain_phragmen {
 			})
 	}
 
+	#[test]
+	fn signed_result_can_be_submitted_regardless_active_era() {
```
Comment (thiolliere) [inline frame/staking/src/tests.rs:2994]: Maybe this test is not very useful, it was in order to test that solution could be submitted for current_era.

File: frame/staking/src/tests.rs (hunk 1 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs:3063]: What would be substantially better is to change the test setup to also have this one session drift. This is probably a lot of work, but I think a good step forward. In that case, I would expect active_era and current_era difference to matter. 

Also related.

File: frame/staking/src/mock.rs [hunk 3]
```diff
@@ -505,6 +505,10 @@ pub fn active_era() -> EraIndex {
 	Staking::active_era().unwrap().index
 }
 
+pub fn current_era() -> EraIndex {
```
Comment (kianenigma) [inline frame/staking/src/mock.rs:508]: given your description: 

> The active era should only be used for rewards stuff.

I think you can now remove all occurrences of acrive_era() and remove the function up there ^^.

File: frame/staking/src/tests.rs (hunk 2 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs:2994]: I think it is not necessary. 

What matters more is a test to ensure if the era that comes with the solution != current_era, then we reject. 

But I think we already have that.

File: frame/staking/src/tests.rs (hunk 1 above)
Comment (thiolliere) [reply frame/staking/src/tests.rs:3063]: indeed if we want that we need to use a different session-pallet which have more than one session delay.

File: frame/staking/src/mock.rs [hunk 4]
```diff
@@ -612,9 +612,13 @@ pub fn start_session(session_index: SessionIndex) {
 	assert_eq!(Session::current_index(), session_index);
 }
 
+// This start and activate the era given.
+// Because the mock use pallet-session which delays session by one, this will be one session after
+// the election happened, not the first session after the election has happened.
 pub fn start_era(era_index: EraIndex) {
```
Comment (thiolliere) [inline frame/staking/src/mock.rs:618]: I modified the test to use current_era instead of active_era but actually the tests generally as you pointed in https://github.com/paritytech/substrate/issues/5244 the tests doesn't test this difference.

Also start_era is ambiguous and used to both start the current era or the active era.
But in the implementation it actually starts the active_era (without taking into account any Forced stuff, maybe would be better to change it with a while loop.

File: frame/staking/src/offchain_election.rs [hunk 5]
```diff
@@ -113,15 +113,15 @@ pub(crate) fn compute_offchain_election<T: Trait>() -> Result<(), OffchainElecti
 	// process and prepare it for submission.
 	let (winners, compact, score) = prepare_submission::<T>(assignments, winners, true)?;
 
-	// defensive-only: active era can never be none except genesis.
-	let era = <Module<T>>::active_era().map(|e| e.index).unwrap_or_default();
+	// defensive-only: current era can never be none except genesis.
+	let current_era = <Module<T>>::current_era().unwrap_or_default();
```
Comment (thiolliere) [inline frame/staking/src/offchain_election.rs:117]: Also I don't think this is defensive, if current_era is none it means there is no current_era, maybe the staking module hasn't started yet, but anyway I think it should return OffchainElectionError, but this is out of scope of the PR

**Comment by thiolliere** [comment]
I updated the tests but they still doesn't cover this change indeed. should https://github.com/paritytech/substrate/issues/5244 be part of this PR or should we merge this as is ?

File: frame/staking/src/tests.rs [hunk 6]
```diff
@@ -3780,7 +3780,7 @@ mod offchain_phragmen {
 				run_to_block(12);
 				assert_eq!(Staking::era_election_status(), ElectionStatus::Open(12));
 
-				let offender_expo = Staking::eras_stakers(active_era(), 10);
+				let offender_expo = Staking::eras_stakers(Staking::active_era().unwrap().index, 10);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:3783]: are we sure this is correct and the `eras_stakers` should be indexed by active era here?

Comment (thiolliere) [reply frame/staking/src/tests.rs:3783]: offence can come from active era or previous era (i.e. a validator must be validating or having been validating in order to commit an offence), it reports the exposures of the era of the offence.

Comment (kianenigma) [reply frame/staking/src/tests.rs:3732]: it depends on this:

```rust
pub(crate) fn on_offence_now(
	offenders: &[OffenceDetails<AccountId, pallet_session::historical::IdentificationTuple<Test>>],
	slash_fraction: &[Perbill],
) {
	let now = Staking::active_era().unwrap().index;
	on_offence_in_era(offenders, slash_fraction, now)
}
```

it is fine for now.
